subprojects/messaging/src/main/java/org/gradle/internal/serialize/ExceptionPlaceholder.java

```diff
@@ -142,20 +142,53 @@ public Object transform(Object obj) {
         return Collections.emptyList();
     }
 
-    private static List<ExceptionPlaceholder> convertToExceptionPlaceholderList(List<? extends Throwable> throwables, Transformer<ExceptionReplacingObjectOutputStream, OutputStream> objectOutputStreamCreator) {
+    private static List<ExceptionPlaceholder> convertToExceptionPlaceholderList(Throwable original, List<? extends Throwable> throwables, Transformer<ExceptionReplacingObjectOutputStream, OutputStream> objectOutputStreamCreator, boolean checkCycle) {
         if (throwables.isEmpty()) {
             return Collections.emptyList();
         } else if (throwables.size() == 1) {
-            return Collections.singletonList(new ExceptionPlaceholder(throwables.get(0), objectOutputStreamCreator));
+            Throwable cause = throwables.get(0);
+            if (checkCycle) {
+                if (hasCycle(original, cause)) {
+                    return Collections.emptyList();
+                }
+            }
+            return Collections.singletonList(new ExceptionPlaceholder(cause, objectOutputStreamCreator));
         } else {
             List<ExceptionPlaceholder> placeholders = new ArrayList<ExceptionPlaceholder>(throwables.size());
             for (Throwable cause : throwables) {
-                placeholders.add(new ExceptionPlaceholder(cause, objectOutputStreamCreator));
+                if (!checkCycle || !hasCycle(original, cause)) {
```

Comment:
So `hasCycle` walks the whole chain of `suppressedExceptions` for `cause`. That essentially means when we now construct an `ExceptionHolder` for `cause`, we'll revisit the same chain just one level below again.

Reply:
Right, going to refactor that with passing down a `Set` containing all seen exceptions.